tf_encrypted/keras/layers/dense.py

```diff
@@ -0,0 +1,121 @@
+  (only applicable if `use_bias` is `True`).
+
+  Arguments:
+      units: Positive integer, dimensionality of the output space.
+      activation: Activation function to use.
+          If you don't specify anything, no activation is applied
+          (ie. "linear" activation: `a(x) = x`).
+      use_bias: Boolean, whether the layer uses a bias vector.
+      kernel_initializer: Initializer for the `kernel` weights matrix.
+      bias_initializer: Initializer for the bias vector.
+      kernel_regularizer: Regularizer function applied to
+          the `kernel` weights matrix.
+      bias_regularizer: Regularizer function applied to the bias vector.
+      activity_regularizer: Regularizer function applied to
+          the output of the layer (its "activation")..
+      kernel_constraint: Constraint function applied to
+          the `kernel` weights matrix.
+      bias_constraint: Constraint function applied to the bias vector.
+
+  Input shape:
+      nD tensor with shape: `(batch_size, ..., input_dim)`.
+      The most common situation would be
+      a 2D input with shape `(batch_size, input_dim)`.
+
+  Output shape:
+      nD tensor with shape: `(batch_size, ..., units)`.
+      For instance, for a 2D input with shape `(batch_size, input_dim)`,
+      the output would have shape `(batch_size, units)`.
+  """
+
+  def __init__(self,
+               units,
+               activation=None,
+               use_bias=True,
+               kernel_initializer='glorot_uniform',
+               bias_initializer='zeros',
+               kernel_regularizer=None,
+               bias_regularizer=None,
+               activity_regularizer=None,
+               kernel_constraint=None,
+               bias_constraint=None,
+               **kwargs):
+
+    super(Dense, self).__init__(**kwargs)
+
+    self.units = int(units)
+    self.activation = activations.get(activation)
+    self.use_bias = use_bias
+
+    self.kernel_initializer = initializers.get(kernel_initializer)
+    self.bias_initializer = initializers.get(bias_initializer)
+
+    if kernel_regularizer:
+      raise NotImplementedError(argument_msg.format("kernel_regularizer",
+                                                    "Dense"))
+    if bias_regularizer:
+      raise NotImplementedError(argument_msg.format("bias_regularizer",
+                                                    "Dense"))
+    if activity_regularizer:
+      raise NotImplementedError(argument_msg.format("activity_regularizer",
+                                                    "Dense"))
+    if kernel_constraint:
+      raise NotImplementedError(argument_msg.format("kernel_constraint",
+                                                    "Dense"))
+    if bias_constraint:
+      raise NotImplementedError(argument_msg.format(" bias_constraint",
+                                                    "Dense"))
+
+  def compute_output_shape(self, input_shape):
+    return [input_shape[0] + self.units]
```

Comment:
is this consistent with https://github.com/tf-encrypted/tf-encrypted/pull/481/files#diff-a4f929bc8ab0de50f08cd8542bd67801R46 and https://github.com/tf-encrypted/tf-encrypted/pull/481/files#diff-a4f929bc8ab0de50f08cd8542bd67801R51 ?

Reply:
I think it is. But it needs a coma instead of a + sign.

Reply:
does it not limit the output shape to two dimension?

Reply:
ah yes, you are right. I guess we could update the comment to 2D tensor (as an input and output) for the Dense. No sure we want to go beyond 2d tensors for now for the dense layer.

Reply:
It does, and it's on purpose.  Dense layers should only output rank 2 tensors according to the Keras spec.  Gemm is not implemented in Keras afaik

Reply:
@jvmancuso do you have an opinion regarding this?

Reply:
From [here](https://www.tensorflow.org/api_docs/python/tf/keras/layers/Dense):

> Note: if the input to the layer has a rank greater than 2, then it is flattened prior to the initial dot product with kernel.

Is it worth adding that logic?

Reply:
I normally do the Flatten explicitly

Reply:
but if that's how they do it we should aim for the same

Reply:
Same here. And in the official Keras examples, they also do Flatten explicitly:
https://github.com/keras-team/keras/blob/master/examples/mnist_cnn.py

Reply:
there's a bit more going on: https://github.com/tensorflow/tensorflow/blob/r1.13/tensorflow/python/keras/layers/core.py#L966-L973

if you guys don't think people will fall into to (including when they load models) then I'm okay to skip

Reply:
ah, I missed this comment.

Reply:
I think if tf.keras does it we should follow suit.  it should be an easy addition.